### PR TITLE
Propagate CreationMeta when chaining views

### LINKED ISSUE #49824: View CreationMeta are not propagated when chaining views
As per title.
Example failing code that generates potentially wrong gradients:

```python
import torch

a = torch.rand(10, requires_grad=True).clone()

# View for which inplace are invalid
b = a.unbind(0)
# Chain of view where the last one is valid for inplace
c = b[0].view_as(b[0])

# This one should fail but does not
c.mul_(2)
# This one properly fails
b.mul_(2)
```

In general, the CreationMeta should be inherited from the previous view to ensure invalid chain of views are properly forbidden.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

### PR DESCRIPTION
**Pull request by jbschlosser**
Fixes #49824

## Background

When creating a view of a view, there was a possibility that the new view would be less restrictive than the previous view, incorrectly sidestepping the error that should be thrown when using in-place operations on the new view.

The fix addresses this by propagating `CreationMeta` from the previous view to the new view. Currently, the old view's `creation_meta` is only propagated when the new view's `creation_meta == CreationMeta::DEFAULT`. This ensures that the new view is not less restrictive than the previous view wrt. allowing in-place operations.

## Test Plan
```
python test/test_autograd.py TestAutogradDeviceTypeCPU.test_inplace_view_of_multiple_output_view_cpu
python test/test_autograd.py TestAutogradDeviceTypeCUDA.test_inplace_view_of_multiple_output_view_cuda
python test/test_autograd.py TestAutogradDeviceTypeCPU.test_inplace_multiple_output_view_of_view_cpu
python test/test_autograd.py TestAutogradDeviceTypeCUDA.test_inplace_multiple_output_view_of_view_cuda
```

## BC-Breaking Note
After this fix, an error will properly be thrown to avoid wrong gradients when an in-place operation is performed on a view of a view, where the second view has `creation_meta == CreationMeta::DEFAULT` (e.g. a standard single output view) while the first view has `creation_meta != CreationMeta::DEFAULT` (e.g. a multiple output or other restricted view).

Since previous behavior allowed such in-place operations without an error, this change is BC-breaking.

### CODE REVIEW

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 1]
```diff
@@ -182,6 +184,7 @@ inline Tensor as_view(const Tensor & base, const Tensor & tensor, bool is_bw_dif
       auto diff_view_meta = static_cast<DifferentiableViewMeta*>(base_meta);
       const auto& base_fw_info = diff_view_meta->get_forward_view();
       new_fw_info = base_fw_info.chain(base, tensor, view_func);
+      creation_meta = propagate_creation_meta(diff_view_meta->get_creation_meta(), creation_meta);
```
Comment (albanD) [inline torch/csrc/autograd/VariableTypeUtils.h:187]: I don't think you need to replicate this and l 170.
You can just add this below just next to the call to `make_variable_differentiable_view`.

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 2]
```diff
@@ -229,6 +233,7 @@ inline std::vector<Tensor> as_view(const Tensor & base, std::vector<Tensor>& ten
       // It is ok to create a ViewInfo where only the base is correct in this case as inplace operations on such views are
       // not allowed
       new_fw_info = ViewInfo(base_fw_info.base_, /* view_func */ nullptr);
+      creation_meta = propagate_creation_meta(diff_view_meta->get_creation_meta(), creation_meta);
```
Comment (albanD) [inline torch/csrc/autograd/VariableTypeUtils.h:236]: Same comment about deduplication here.

**Review by albanD — APPROVED** [review]
LGTM!
